src/meta/src/stream/stream_manager.rs

```diff
@@ -104,6 +106,116 @@ where
         })
     }
 
+    async fn resolve_chain_node(
+        &self,
+        table_fragments: &mut TableFragments,
+        ctx: &mut CreateMaterializedViewContext,
+        locations: &ScheduledLocations,
+    ) -> Result<()> {
+        let upstream_parallel_unit_info = self
+            .fragment_manager
+            .get_sink_parallel_unit_ids(ctx.table_sink_map.keys().cloned().collect_vec())
+            .await?;
+
+        for fragment in table_fragments.fragments.values_mut() {
+            // TODO: currently materialize and chain node will be in separate fragments, but they
+            // could be merged into one fragment if they shared the same distribution. We should
+            // also consider FragmentType::Sink once we support merging materialize and
+            // chain into the same fragment.
+            if fragment.fragment_type == FragmentType::Others as i32 {
+                for actor in &mut fragment.actors {
+                    // BFS through all nodes in actor.
+                    let mut queue: VecDeque<&mut StreamNode> = VecDeque::default();
```

Comment:
would you please explain a little bit why we need a BFS here? I think it's easier to understand if we write it recursively... as in other places.

Reply:
Please try to simplify this function, like moving the inner loop into another function, etc.

Reply:
+1 for recursively for better readable.

Reply:
`resolve_chain_node` just simply search for chain nodes in an actor, and BFS comes to my mind at first... 😅 

> I think it's easier to understand if we write it recursively... as in other places.

I agree, let me refactor that.